netket/optimizer/qgt/qgt_jacobian_pytree_logic.py

```diff
@@ -0,0 +1,226 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, Tuple
+from functools import partial
+
+import jax
+import jax.flatten_util
+import jax.numpy as jnp
+
+import numpy as np
+
+from netket.stats import sum_inplace, subtract_mean
+from netket.utils import n_nodes
+
+from netket.utils.types import Array, Callable, PyTree, Scalar
+
+import netket.jax as nkjax
+from netket.jax import tree_cast, tree_conj, tree_axpy, tree_to_real
+
+
+def sub_mean(oks: PyTree) -> PyTree:
+    return jax.tree_map(partial(subtract_mean, axis=0), oks)  # MPI
+
+
+@partial(jax.vmap, in_axes=(None, None, 0))
+def vmap_grad_real_holo(forward_fn: Callable, params: PyTree, samples: Array) -> PyTree:
+    """Calculates Jacobian entries by vmapping grad.
+    Assumes the function is R→R or holomorphic C→C, so single grad is enough
+
+    Args:
+        forward_fn: the log wavefunction, partialed with any model state, limited to single sample, taking a split real-imaginary PyTree in R→R mode
+        params : a pytree of parameters p, real-imaginary split in R→R mode
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
+    """
+    y, vjp_fun = jax.vjp(forward_fn, params, samples)
+    res, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+    return res
+
+
+def vmap_grad_centered_real_holo(
+    forward_fn: Callable, params: PyTree, samples: Array
+) -> PyTree:
+    """Calculates centred Jacobian (i.e., subtracts MPI mean from vmap_grad)"""
+    return sub_mean(vmap_grad_rr_cc(forward_fn, params, samples))
```

Comment:
```suggestion
    return sub_mean(vmap_grad_real_holo(forward_fn, params, samples))
```
Was changing the nomenclature too quickly!